### PR TITLE
Fix: enable out_mat_r in nscf calculation

### LINKED ISSUE #4045: Besides data-HR and data-SR, I need an rR file without scf calculation
### Background

rR is the output when I turn on "out_mat_r" tag. 
b6f443f71a7d3f630f2a2357042d787

### Describe the solution you'd like

Is it possible to get the rR matrix without scf calculation?

### PR DESCRIPTION
**Pull request by dyzheng**
### Reminder
- [ ] Have you linked an issue with this pull request?
- [ ] Have you added adequate unit tests and/or case tests for your pull request?
- [ ] Have you noticed possible changes of behavior below or in the linked issue?
- [ ] Have you explained the changes of codes in core modules of ESolver, HSolver, ElecState, Hamilt, Operator or Psi? (ignore if not applicable)

### Linked Issue
Fix #4045 
enable setting “out_mat_r” with "calculation nscf".

### Unit Tests and/or Case Tests for my changes
- A unit test is added for each new feature or bug fix.

### What's changed?
- Example: My changes might affect the performance of the application under certain conditions, and I have tested the impact on various scenarios...

### Any changes of core modules? (ignore if not applicable)
- Example: I have added a new virtual function in the esolver base class in order to ...

### CODE REVIEW

File: source/module_esolver/esolver_ks_lcao_elec.cpp [hunk 1]
```diff
@@ -716,6 +716,9 @@ void ESolver_KS_LCAO<TK, TR>::nscf(void)
         GlobalC::ld.cal_gedm(GlobalC::ucell.nat);
     }
 #endif
+
+    this->create_Output_Mat_Sparse(0).write();
```
Comment (mohanchen) [inline source/module_esolver/esolver_ks_lcao_elec.cpp:714]: any new tests needed for this function?

Comment (dyzheng) [reply source/module_esolver/esolver_ks_lcao_elec.cpp:714]: In SCF calculation, the output of Hamiltonian, Overlap, rR matrix has been supported and tested.
But in NSCF calculation, these files won't be printed with input parameters, which looks like a bug.
Some users want to skip SCF calculation and output these files.
New tests for this feature is necessary, I will modify existed test to cover this feature without extra testing time.

Comment (mohanchen) [reply source/module_esolver/esolver_ks_lcao_elec.cpp:714]: OK， then I will merge this PR after the tests have been added.

Comment (dyzheng) [reply source/module_esolver/esolver_ks_lcao_elec.cpp:714]: There are two bugs discovered by this PR: 
1. `out_mat_dh 1` lead to different results with different MPI cores.
2. `out_mat_dh 1` will affect the result of `out_mat_r 1` .
I add test to cover `out_mat_hs2`, `out_mat_r`, `out_mat_t`, without test of `out_mat_dh`, which I plan to submit another PR to fix it.

Comment (mohanchen) [reply source/module_esolver/esolver_ks_lcao_elec.cpp:714]: OK

**Review by mohanchen — APPROVED** [review]

**Pull request merged**